bin/s3stalker_daemon.py

```diff
@@ -0,0 +1,176 @@
+        self.time_back = self.config['timedelta']
+        self.publisher = None
+        self.loop = False
+        self._setup_and_start_communication()
+
+    def _setup_and_start_communication(self):
+        """Set up the Posttroll communication and start the publisher."""
+        logger.debug("Starting up... ")
+        self.publisher = create_publisher_from_dict_config(self.config['publisher'])
+        self.publisher.start()
+        self.loop = True
+        signal.signal(signal.SIGTERM, self.signal_shutdown)
+
+    def signal_shutdown(self, *args, **kwargs):
+        """Shutdown the S3 Stalker daemon/runner."""
+        self.close()
+
+    def run(self):
+        """Start the s3-stalker daemon/runner in a thread."""
+        first_run = True
+        config = self.config
+        last_fetch_time = None
+        while self.loop:
+            if first_run:
+                logger.info('Create messages with urls for most recent files only')
+                logger.info('On start up we consider files with age up to %d seconds from now',
+                            self.startup_timedelta_seconds)
+                config['timedelta'] = {'seconds': self.startup_timedelta_seconds}
+                first_run = False
+            else:
+                seconds_back = self._get_seconds_back_to_search(last_fetch_time)
+                config['timedelta'] = {'seconds': seconds_back}
+
+            messages = create_messages_for_recent_files(self.bucket, config)
+
+            last_fetch_time = get_last_fetch()
+            logger.info("Last fetch time...: %s", str(last_fetch_time))
+
+            for message in messages:
+                logger.info("Publishing %s", str(message))
+                self.publisher.send(str(message))
+
+            waiting_time = timedelta(**self.time_back)
+            wait_seconds = waiting_time.total_seconds()
+
+            # Wait for some time...
+            logger.debug("Waiting %d seconds", wait_seconds)
+            time.sleep(max(wait_seconds, 0))
+
+    def _get_seconds_back_to_search(self, last_fetch_time):
+        """Update the time to look back considering also the modification time of the last file."""
+        dtime_back = timedelta(**self.time_back)
+        seconds_back = dtime_back.total_seconds()
+
+        if last_fetch_time is None:
+            return seconds_back
+
+        start_time = datetime.utcnow()
+        start_time = start_time.replace(tzinfo=timezone.utc)
+        seconds_to_last_file = (start_time - last_fetch_time).total_seconds()
+        return max(seconds_back, seconds_to_last_file)
+
+    def close(self):
+        """Shutdown the S3Stalker runner."""
+        logger.info('Terminating the S3 Stalker daemon/runner.')
+        self.loop = False
+        if self.publisher:
+            try:
+                self.publisher.stop()
+            except Exception:
+                logger.exception("Couldn't stop publisher.")
```

Comment:
I think this try/except isn't needed, the `if self.publisher` check handles the case where it is `None`, and otherwise it is a safe operation. At least there are no try/except in any of the places it is used I checked :-)

Reply:
Ok, I see. You are probably right.